### ISSUE TITLE
physical keyboard desktop -> mobile

### ISSUE THREAD

**Issue by michalwitek1232** [rustdesk/rustdesk#3556]
### Bug Description

Hi, keyboard used by client does not send keys into remote device.

Also, to be sure, I tested every key mapping from this list:
![obraz](https://user-images.githubusercontent.com/42500203/223693470-50f90710-8528-450e-999a-94abca1177d6.png)

None worked

### How to Reproduce

When I'm pressing any keys on my keyboard, for ex. when trying to type: google.com, when I'm pressing `g o o g l e . c o m`, My android remote does not put those keys into a text field ( in my example, web browser search bar ).

### Expected Behavior

I should be able send my keyboard keys into remote with ease, so I could just type on my keyboard instead of remote virtual keyboard (ex. google keyboard, swiftkeys etc)

### Operating system(s) on local side and remote side

Windows 11 -> Android

### RustDesk Version(s) on local side and remote side

nightly -> nightly (52ce3dd)

### Screenshots

Permissions on remote:
![obraz](https://user-images.githubusercontent.com/42500203/223693003-7b39a340-bb3e-4c09-8720-837150f66c1e.png)

### Additional Context

Android device system info:
POCO F2 Pro, MIUI Global 13.0.3 (Stable)
Android 12 SKQ1.211006.001

Connection by self-hosted relay

**Comment by rustdesk** [comment]
Yes, soft keyboard only for android host.
Close as wontfix

**Comment by rustdesk** [comment]
We can reopen this, if some one can figure out a solution.

**Comment by michalwitek1232** [comment]
btw, any idea why this is not working? I mean any clue?

**Comment by michalwitek1232** [comment]
I read about something called [wifi keyboard](https://f-droid.org/zh_Hans/packages/com.volosyukivan/), maybe it's possible to integrate it with rustdesk. I'll try to do it. Wish me luck @rustdesk

**Comment by rustdesk** [comment]
Interesting

**Comment by rustdesk** [comment]
@Heap-Hop

**Comment by DonRichie** [comment]
> Yes, soft keyboard only for android host. Close as wontfix

I am sorry, but I can't understand this answer. Is there a way to forward keyboard input to a controlled android device?
On my Android 13 device RustDesk is activated as accessibility app, but I can't enter keys, only mouse.
Do I miss something?

Linux -> Android 13

**Comment by Zhurama** [comment]
Rustdesk Mobile really does... need hardware keyboard support. I was fortunate today that my Corsair KB has BT.. but that won't help remote sessions.. I hope you folks can figure it out.. I'm loving the service...

**Comment by xbyte99** [comment]
Somehow the TeamViewer guys made the remote keyboard (PC -> Android) working in their QuickSupport as well as in their Teamviewer Host App. So there's got to be a way.
It's especially bad when screen capturing is blocked (e.g. Password entry on lock screen and other apps). With the remote keyboard it would be at least still possible to type in the password and press enter (works in Teamviewer Quicksupport/Host) but on a blacked out screen with virtual keyboard - no chance.

**Comment by TheGrave** [comment]
> Somehow the TeamViewer guys made the remote keyboard (PC -> Android) working in their QuickSupport as well as in their Teamviewer Host App. So there's got to be a way. It's especially bad when screen capturing is blocked (e.g. Password entry on lock screen and other apps). With the remote keyboard it would be at least still possible to type in the password and press enter (works in Teamviewer Quicksupport/Host) but on a blacked out screen with virtual keyboard - no chance.

On the other hand I can't get an unlock prompt in TV QS/Host on Sony Xperia due a long known issue when Fingerprint unlock is enabled. That's why I decided to give Rustdesk a chance but hitting the same issue with lack of remote keyboard support really puts me off. Super cumbersome to type with a cursor.

**Comment by wrjcs** [comment]
Can’t be resolved?

**Comment by feliciien** [comment]
Hello the bug is only on windows 11 -> android 
Or windows -> android 
Or any desktop (MAC- pc) -> android ?

**Comment by xbyte99** [comment]
I can confirm this bug for windows 10 + 11

**Comment by feliciien** [comment]
Thanks

**Comment by TheGrave** [comment]
> Thanks

Ubuntu 20.04 as well.

**Comment by wrjcs** [comment]
> 💎 **$300** bounty created by rustdesk 🙋 If you start working on this, comment `/attempt #3556` to notify everyone 👉 To claim this bounty, submit a pull request that includes the text `/claim #3556` somewhere in its body 📝 Before proceeding, please make sure you can [**receive payouts**](https://docs.algora.io/bounties/payments#country-support) in your country 💵 Payment arrives in your account 2-5 days after the bounty is rewarded 💯 You keep 100% of the bounty award 🙏 Thank you for contributing to rustdesk/rustdesk!
> 
> Attempt	Started (GMT+0)	Solution
> 🟢 @feliciien	Sep 30, 2023, 10:58:35 PM	WIP

I can pay $1000 for this issue

**Comment by rustdesk** [comment]
> I can pay $1000 for this issue

How to pay?

**Comment by wrjcs** [comment]
Paypal or other method?

**Comment by rustdesk** [comment]
> Paypal or other method?

Great, you can pay to @feliciien directly once his PR is approved.

**Comment by wrjcs** [comment]
No problem

**Comment by algora-pbc** [comment]
@wrjcs would you like to add your bounty into the pool?
👉 https://algora.io/org/rustdesk/bounties/new

**Comment by feliciien** [comment]
Great I have PayPal too 
I'm still working on this issue 📆⏳

**Comment by wrjcs** [comment]
> @wrjcs would you like to add your bounty into the pool? 👉 https://algora.io/org/rustdesk/bounties/new

How to add into the pool?

**Comment by algora-pbc** [comment]
💎 **$1,000** bounty has been added to pool by wrjcs!

**Comment by feliciien** [comment]
Still working on it

**Comment by feliciien** [comment]
I would like to withdraw myself from this issue

**Comment by rustdesk** [comment]
> I would like to withdraw myself from this issue

It is fine, since macfans almost finish this.

**Comment by michalwitek1232** [comment]
I'm glad that someone took care about this issue.

**Comment by fragtion** [comment]
Were there any other changes to the the Android/aarch64 apk (besides for this new feature) since 2023/11/07 build?

2023/11/08 build launches to a blank/black screen when opening the RustDesk app:

rustdesk-1.2.4-aarch64-signed.apk 2023/11/08 build (cffa3943aa8ea6ee7e53f262abc2f2f9):
![WhatsApp Image 2023-11-08 at 13 15 18_067350ac](https://github.com/rustdesk/rustdesk/assets/2876889/a9567c53-3c15-4a39-8bab-4a466f86e703)

I tried to uninstall and reinstall the app (Wipe app Data & Cache) but problem persists.

2023/11/07 build worked fine so it's some change since yesterday that caused this

Galaxy note 8 with DeluxeROM (Android version 9, One UI version 1.0, Kernel version 4.4.111++DeluxeKernel_v17.1++)

**Comment by wrjcs** [comment]
Hello @mcfans ,I have paid for this issue.
![image](https://github.com/rustdesk/rustdesk/assets/7202194/7c5db231-ef43-4d7e-a135-83e0d3ddde23)

**Comment by mcfans** [comment]
Hi @fragtion! I reproduced the same issue on my side. I got these logs from logcat.
<img width="1414" alt="image" src="https://github.com/rustdesk/rustdesk/assets/15027395/4137a1f0-4775-4d42-84b7-846991c96850">
I think it might be related with [this commit](https://github.com/rustdesk/rustdesk/commit/8bf57f1293abee463a935aae58f84d01b60f9b74#diff-6670c5314a2ea23963709506313c74a348f27bd18831534b16621a44b6771696L77).

**Comment by fragtion** [comment]
> > Hi @fragtion! I reproduced the same issue on my side. I got these logs from logcat. <img alt="image" width="1414" src="https://user-images.githubusercontent.com/15027395/281479168-4137a1f0-4775-4d42-84b7-846991c96850.png"> I think it might be related with [this commit](https://github.com/rustdesk/rustdesk/commit/8bf57f1293abee463a935aae58f84d01b60f9b74#diff-6670c5314a2ea23963709506313c74a348f27bd18831534b16621a44b6771696L77).
> 
> yes, fixed

Now this latest nightly is crashing for me on launch. "RustDesk keeps stopping"

Logcat:
`
11-09 17:06:51.800 20878 20901 W linker  : Warning: "/data/app/com.carriez.flutter_hbb-SjyrFenMUxL-gImT3O202g==/lib/arm64/librustdesk.so" has unsupported flags DT_FLAGS_1=0x81 (ignoring unsupported flags)
11-09 17:06:51.952 20941 20941 F DEBUG   : Abort message: 'called `Result::unwrap()` on an `Err` value: GlobError { path: "/storage/emulated/0/RustDesk/Logs", error: Os { code: 13, kind: PermissionDenied, message: "Permission denied" } }'
11-09 17:06:51.953 20941 20941 F DEBUG   :     #01 pc 00000000010d6ecc  /data/app/com.carriez.flutter_hbb-SjyrFenMUxL-gImT3O202g==/lib/arm64/librustdesk.so (offset 0x55f000)
`

**Comment by rustdesk** [comment]
> **Update: fixed it by manually removing "/storage/emulated/0/RustDesk/" folder**

@21pages follow up this

**Comment by basilgello** [comment]
How do you get `/storage/emulated/0/RustDesk` folder created? I dont have it on all my devices.

**Comment by rustdesk** [comment]
> /storage/emulated/0

@21pages need to check if `/storage/emulated/0` exists. also find out the unwrap.

**Comment by 21pages** [comment]
@fragtion If you can reproduce the error, could you try updating flexi_logger version to 0.27.3?

**Comment by wrjcs** [comment]
when can we have this function?

**Comment by fragtion** [comment]
> when can we have this function?

The keyboard functionality for Android is already merged but you'll need to upgrade to the nightly version to use it as the last stable public release of RustDesk was released before this feature was added.
You can get the nightly here https://github.com/rustdesk/rustdesk/releases/tag/nightly

**Comment by TheGrave** [comment]
> > when can we have this function?
> 
> The keyboard functionality for Android is already merged but you'll need to upgrade to the nightly version (v1.2.4) to use it as the last stable public version of RustDesk (v1.2.3) was released before this feature was added. You can get the nightly here https://github.com/rustdesk/rustdesk/releases/tag/nightly

Nightly is currently broken:

https://github.com/rustdesk/rustdesk/issues/6360

**Comment by Zahrun** [comment]
Nightly is not broken anymore.
I tested this feature. It works, but it seems that it assumes the keyboard layout to be qwerty, whereas I am using something different. Also, it does not help to unlock the screen on Android side.

**Comment by Zahrun** [comment]
Which one, the keyboard layout or the unlock? By unlock I meant enter the android password. Maybe I was unclear.

**Comment by TheGrave** [comment]
Unlocking has always worked. This is the main reason I started using RustDesk - you can unlock PIN or pattern, can't do it with anything else on Android and I tested them all, believe me. Problem is that stupid service crashes from time to time (or disables input control for no reason) so I ended up using TeamViewer Host as a backup when I need to re-enable it and I had to get rid of the screen lock after all. There is no single remote access app that is reliable enough to do it all on Android at present.

23 Nov 2023 10:22:43 Aroun ***@***.***>:

**Comment by rustdesk** [comment]
> Problem is that stupid service crashes from time to time

How often? https://github.com/rustdesk/rustdesk/issues/6216

**Comment by rustdesk** [comment]
> you can unlock PIN or pattern

Unlock work for you? Actually we do not enable unlock on android, because it does not work in our test.

https://github.com/rustdesk/rustdesk/blob/34fa82f72c6b79fd106602ed375e1fe9e836348d/src/server/video_service.rs#L367

**Comment by TheGrave** [comment]
Dunno, I haven't timed it :) Almost every time I use it more than 20-30 mins.

**Comment by TheGrave** [comment]
> > you can unlock PIN or pattern
> 
> Unlock work for you? Actually we do not enable unlock on android, because it does not work in our test.
> 
> https://github.com/rustdesk/rustdesk/blob/34fa82f72c6b79fd106602ed375e1fe9e836348d/src/server/video_service.rs#L367

That's wierd. On Xperia 1 and 1 III it works fine.

**Comment by Zahrun** [comment]
My experience is that through rustdesk I can wake the phone, but then the PIN screen is hidden, it is all black. If I press the correct places on this black screen where the numbers should be shown, the phone unlocks. This behavior has been described in https://github.com/rustdesk/rustdesk/issues/816.

**Comment by basilgello** [comment]
Works for me: Android 7 (LOS 14.1) -> Android 9-14

**Comment by TheGrave** [comment]
> > Problem is that stupid service crashes from time to time
> 
> How often? #6216

Every time I use it for a while. 30 mins enough usually. Initially I thought it's a conflict with TeamViewer Host running at the same time (without screen recording launched from the latter). I tried removing Host app and RustDesk still crashes on its own.
